lua_modules/websocket/websocket.lua

```diff
@@ -0,0 +1,496 @@
+      if string.match(c, "HTTP/1.1 101(.*)\r\n\r\n") then 
+        print("Websocket found hdr")
+        isHdrRecvd = true
+        m.isConnected = true
+        if m.onConnectedCallback ~= nil then 
+          node.task.post(node.task.LOW_PRIORITY, function()
+            m.onConnectedCallback(host, port, path) 
+          end)
+        end 
+      end 
+    else 
+      -- we can start to deliver incoming data 
+  
+      m.decodeFrame(c)
+    end 
+    
+  end)
+  sk:on("sent", function()
+    -- print("Websocket sent data")
+  end)
+  sk:on('disconnection', function(errcode)
+    print("Websocket disconnection. err:", errcode)
+    m.isConnected = false
+    if m.onDisconnectCallback then m.onDisconnectCallback() end 
+  end)
+  sk:on('reconnection', function(errcode)
+    print('Websocket reconnection. err:', errcode)
+  end)
+  sk:on("connection", function(sck, c)
+    -- print("Websocket got connection. Sending TCP msg. body:")
+    print(body)
+    -- m.sock = sck 
+    sck:send(body)
+    
+    m.pingStart()
+  end)
+  
+  m.sock = sk
+  sk:connect(port, host)
+  
+end
+
+-- currently only supports minimal frame decode
+m.lenExpected = 0
+m.buffer = ""
+-- private method
+function m.decodeFrame(frame)
+  
+  local data, fin, opcode
+  
+  if m.lenExpected > 0 then 
+    -- we should just raw append
+    data = frame 
+    -- print("Raw len:", string.len(data))
+    -- m.lenExpected = m.lenExpected - string.len(data)
+    
+    -- append the new data to the previous data
+    m.buffer = m.buffer .. data
+
+  else
+    -- we need to decode the headr from the frame 
+  
+    -- get FIN. 1 means msg is complete. 0 means multi-part
+    fin = string.byte(frame, 1)
+    fin = bit.isset(fin, 7)
+    -- print("FIN:", fin)
+    
+    -- get opcode
+    opcode = string.byte(frame, 1)
+    opcode = bit.clear(opcode, 4, 5, 6, 7) -- clear FIN and RSV
+    -- print("Opcode:", opcode)
+    
+
+    -- get 2nd byte as it has the payload length
+    -- msb of byte is mask, remaining 7 bytes is len 
+    local plen = string.byte(frame, 2)
+    local mask = bit.isset(plen, 7)
+    plen = bit.clear(plen, 7) -- remove the mask from the length
+    -- print("Frame len:", plen, " Mask:", mask)
+    m.lenExpected = plen
+    
+    if mask then 
+      -- print("We should not get a mask from server. Error.")
+      return
+    end
+    
+    data = string.sub(frame, 3) -- remove first 2 bytes, i.e. start at 3rd char
+  
+    if plen == 126 then 
+      -- read next 2 bytes
+      local extlen = string.byte(frame, 3)
+      local extlen2 = string.byte(frame, 4)
+      -- bit shift by one byte extlen 
+      extlen = bit.lshift(extlen, 8)
+      local actualLen = extlen + extlen2
+      -- print("ActualLen:", actualLen)
+      m.lenExpected = actualLen
+      
+      data = string.sub(data, 3) -- remove first 2 bytes
+      
+    elseif plen == 127 then 
+      -- print("Websocket lib does not support longer payloads yet")
+      -- return
```

Comment:
Maybe throw an error here? Else finding the problem if it happens could be ugly

Reply:
Will fix that. I'll add error to the url parse too.